### PR TITLE
Refactor unittests for activation functions relu, elu, and sigmoid

### PR DESCRIPTION
**Pull request by paulshaoyuqiao**
Summary: The tests covered previously by test_qrelu, test_qelu, and test_qsigmoid are now merged into one test to ensure conciseness and reduce redundancy.

Test Plan:
1. On a devserver, build PyTorch from source by running the command "buck build mode/dev //caffe2:torch"
 2. Run the merged unit test throught the command "buck test mode/dev //caffe2/test:quantization -- test_qrelu_elu_sigmoid

Differential Revision: D21755690

### CODE REVIEW

**Comment by vkuzo** [comment]
would it be possible to share the code but still have separate tests per activation function?  That way, if one of the activation tests is flaky / disabled / etc, it wouldn't impact the other tests

**Comment by vkuzo** [comment]
> would it be possible to share the code but still have separate tests per activation function? That way, if one of the activation tests is flaky / disabled / etc, it wouldn't impact the other tests

it could be something like 

```
activation_test_config = { ... }

def _test_activation(...):
  ...the shared test logic...

# one test case for each activation
test_foo(self, ...):
    foo_config = activation_test_config['foo']
    _test_activation(foo_config)

```

File: test/quantization/test_quantized_op.py [hunk 1]
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
...
+                    torch.nn.quantized.ReLU6()
+                ],
+                'reference_fn': torch.nn.ReLU6(),
+            },
+            {
+                'quantized_fn': [
+                    torch.nn.quantized.ReLU6(inplace=True)
+                ],
+                'reference_fn': torch.nn.ReLU6(),
+            }
```
Comment (z-a-f) [inline test/quantization/test_quantized_op.py:186]: nit: It looks to me this configurations are very similar. Can you collapse them into one, and just add all of them to the list of `quantized_fn`?

```python
           {
                'quantized_fn': [
                    torch.ops.quantized.relu6,
                    torch.nn.quantized.ReLU6(inplace=False),
                    torch.nn.quantized.ReLU6(inplace=True)
                ],
                'reference_fn': torch.nn.functional.relu6
            },
```

File: test/quantization/test_quantized_op.py [hunk 2]
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
...
+                    torch.relu
+                ],
+                'reference_fn': torch.relu
+            },
+            {
+                'quantized_fn': [
+                    torch.relu_
+                ],
+                'reference_fn': torch.relu_,
+            },
```
Comment (z-a-f) [inline test/quantization/test_quantized_op.py:149]: nit: Can you also merge these two configurations (see example below).

File: test/quantization/test_quantized_op.py [hunk 3]
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
...
+    # quantized_fn: a list of the quantized functions to be tested
+    # reference_fn: the original reference function to be called on the
+    # the dequantized X
+    # inplace_kwarg: the additional inplace keyword argument to test in-place
+    # for each test entry in ops_under_test, it must have at least the fields
+    # for quantized_fn and reference_fn. If inplace_kwarg is missing, the
+    # quantized function is assumed to be either inplace by default or the
+    # test is not testing an inplace function.
+    # output_range: the output range the operator will map to. By default, if it is
+    # no specified, the range will be (-inf, inf)
```
Comment (z-a-f) [inline test/quantization/test_quantized_op.py:135]: nit: Is the range [-inf, inf] or is it dependent on the X min max?

**Review by z-a-f — APPROVED** [review]
LGTM, less some nits

File: test/quantization/test_quantized_op.py [hunk 4]
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
...
+                    # have a constrained output range. e.x. sigmoid ranges from 0 to 1.
+                    output_scale = scale
+                    if 'output_range' in op_group:
+                        (f_min, f_max) = op_group['output_range']
+                        output_scale = (f_max - f_min) / (q_max - q_min + 1.0)
+
+                    # Adjusts output_zero_point if needed (for sigmoid and hardsigmoid).
+                    # output_zero_point determines the additional offset that will be
+                    # added to a scaled value during quantization.
+                    if fn_name in ['sigmoid', 'hardsigmoid']:
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:243]: can you make this a config as well?

File: test/quantization/test_quantized_op.py [hunk 5]
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
...
+    # reference_fn: the original reference function to be called on the
+    # the dequantized X
+    # inplace_kwarg: the additional inplace keyword argument to test in-place
+    # for each test entry in ops_under_test, it must have at least the fields
+    # for quantized_fn and reference_fn. If inplace_kwarg is missing, the
+    # quantized function is assumed to be either inplace by default or the
+    # test is not testing an inplace function.
+    # output_range: the output range the operator will map to. By default, if it is
+    # no specified, the range will be (-inf, inf)
+    activation_test_config = {
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:136]: I feel moving the config to the tests would be better?

Comment (z-a-f) [reply test/quantization/test_quantized_op.py:136]: Do you mean make it part of the helper function below (`_test_activation_function`)? If so, we were originally planning on doing it, but I think it just bloats thee method. Also, we were thinking in the future we can separate the activations functions into separate class, say `TestQuantizedActivationsOps`, which would be configuration based only. In leu of the that, we kinda converged to the class-level config. Do you think we should move it inside the helper method still?

Comment (jerryzh168) [reply test/quantization/test_quantized_op.py:136]: I mean put this in individual tests, like `test_qrelu`, that way it is easier to see what is been tested from the test itself

Comment (jerryzh168) [reply test/quantization/test_quantized_op.py:140]: I think putting in individual tests would mean one less lookup, not sure what is the benefit of putting this in class level?

Comment (paulshaoyuqiao) [reply test/quantization/test_quantized_op.py:140]: That makes sense. By distributing the config individually into each of the test function, they can be changed more easily and directly. I will add that change!

File: test/quantization/test_quantized_op.py [hunk 6]
```diff
@@ -118,68 +118,164 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
...
-                                       dtype=torch_type)
+        # Quantizes the reference to account for max error.
+        # q_min and q_max only depend on the initial torch_type.
+        q_min, q_max = torch.iinfo(torch_type).min, torch.iinfo(torch_type).max
 
-        ops_under_test = {
-            'native': torch.relu,
-            'nn.functional': torch.nn.functional.relu,
-        }
+        op_groups = test_configs
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:158]: nit: remove `test_configs` or rename `op_groups` to `test_config`

File: test/quantization/test_quantized_op.py [hunk 7]
```diff
@@ -118,68 +118,164 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
 
 class TestQuantizedOps(TestCase):
 
-    """Tests the correctness of the quantized::relu op."""
-    @given(X=hu.tensor(shapes=hu.array_shapes(1, 5, 1, 5),
-                       qparams=hu.qparams()))
-    def test_qrelu(self, X):
+    # When writing a unit test for the activation function,
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:121]: nit: use doc string for this comment? and move this to function `_test_activation_function`?

**Review by jerryzh168 — APPROVED** [review]
LGTM, had a few nit comments
